File: legacy/packages/contracts/python/src/gta_ai_contracts/jobs.py
from __future__ import annotations

from datetime import datetime
from typing import Literal, TypeAlias
from uuid import UUID

from pydantic import Field, model_validator

from .common import JsonValue, NonNegativeInt, Sha256, StudioModel, UnitScore

JobStatus: TypeAlias = Literal["QUEUED", "BLOCKED", "LEASED", "RUNNING", "RETRY_WAIT", "SUCCEEDED", "FAILED", "CANCELLED"]


class JobError(StudioModel):
    code: str
    message: str
    retryable: bool
    details: dict[str, JsonValue] = Field(default_factory=dict)


class JobRun(StudioModel):
    id: UUID
    project_id: UUID
    kind: str
    status: JobStatus
    priority: int = 0
    idempotency_key: str
    input_fingerprint: Sha256
    algorithm_version: str
    parameters: dict[str, JsonValue] = Field(default_factory=dict)
    attempt: NonNegativeInt = 0
    max_attempts: int = Field(default=3, gt=0)
    progress: UnitScore = 0
    lease_owner: str | None = None
    lease_expires_at: datetime | None = None
    next_retry_at: datetime | None = None
    cancel_requested_at: datetime | None = None
    error: JobError | None = None
    result_artifact_id: UUID | None = None
    created_at: datetime
    updated_at: datetime

    @model_validator(mode="after")
    def validate_status_fields(self) -> "JobRun":
        if self.status in {"LEASED", "RUNNING"} and (self.lease_owner is None or self.lease_expires_at is None):
            raise ValueError("leased/running jobs require lease owner and expiry")
        if self.status == "RETRY_WAIT" and self.next_retry_at is None:
            raise ValueError("retry jobs require next_retry_at")
        if self.status == "SUCCEEDED" and self.result_artifact_id is None:
            raise ValueError("succeeded jobs require a result artifact")
        return self

File: legacy/packages/contracts/src/jobs.ts
import type { IsoDateTime, JsonObject, Sha256, Uuid } from "./common.js";

export const JOB_STATUSES = [
  "QUEUED",
  "BLOCKED",
  "LEASED",
  "RUNNING",
  "RETRY_WAIT",
  "SUCCEEDED",
  "FAILED",
  "CANCELLED",
] as const;

export type JobStatus = (typeof JOB_STATUSES)[number];

export interface JobError {
  readonly code: string;
  readonly message: string;
  readonly retryable: boolean;
  readonly details: JsonObject;
}

export interface JobRun {
  readonly id: Uuid;
  readonly project_id: Uuid;
  readonly kind: string;
  readonly status: JobStatus;
  readonly priority: number;
  readonly idempotency_key: string;
  readonly input_fingerprint: Sha256;
  readonly algorithm_version: string;
  readonly parameters: JsonObject;
  readonly attempt: number;
  readonly max_attempts: number;
  readonly progress: number;
  readonly lease_owner: string | null;
  readonly lease_expires_at: IsoDateTime | null;
  readonly next_retry_at: IsoDateTime | null;
  readonly cancel_requested_at: IsoDateTime | null;
  readonly error: JobError | null;
  readonly result_artifact_id: Uuid | null;
  readonly created_at: IsoDateTime;
  readonly updated_at: IsoDateTime;
}

